Zork Grand Inquisitor: Implement New Game

nbrochu opened the pull request:
## What is this fixing or adding?
Adds Archipelago support for Zork Grand Inquisitor, the 1997 point-and-click PC adventure game.

The client (based on `CommonClient`), on top of its regular Archipelago duties, fully handles the randomization of the game and the monitoring / modification of the game state. No game modding needed at all; the player is ready to play an Archipelago seed if they can play the vanilla game through ScummVM.

The "reverse engineering" (there's likely a better term for this...) of the game is my own original work and I included an MIT license at the root of my world directory.

A PopTracker pack was also created to help people learn the game: https://github.com/SerpentAI/ZorkGrandInquisitorAPTracker

I realize this is a niche game that not everyone will have the means to test, so I made a video where I play a seed to goal to demonstrate this implementation in action: https://www.youtube.com/watch?v=EXb0QeOU6Wg

## How was this tested?
* A generation stress-test (generates really fast; logic is relatively simple)
* About thirty solo seeds completed
* A handful of larger asyncs completed. Tested with all officially supported Archipelago games
* APWorlds have been available for testing in the [Discord thread](https://discord.com/channels/731205301247803413/1173652682536394783) in #future-game-design
* Played multiple times in scheduled unsupported games on the official Discord server
* Fairly comprehensive unit test coverage around logic and options

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/e4df38ae-516c-4123-b18d-3fda612c7d2c)

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/72b55ae2-d560-4451-b3e9-ca7abb1b4e10)

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/5e227672-269c-4a4d-95c5-2c549715312d)

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/a9dfed66-5a8a-46ed-b913-d9268bbb2662)

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/7dce8b16-33be-4b45-9ad7-4b4ceb0c1fc1)

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/27f11a55-7bcd-4bdd-90ce-80be85b92606)

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/b795c1ff-fcfb-40a5-aff2-7b18107efcab)

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/e6c346ce-ce05-42e4-8744-9ed05a3b4279)

nicholassaylor commented on worlds/zork_grand_inquisitor/docs/setup_en.md:
```diff
@@ -0,0 +1,42 @@
+## Requirements
+
+- Windows OS (Hard required. Client is using memory reading / writing through Win32 API)
+- A copy of Zork Grand Inquisitor. Only the GOG version is supported. The Steam version can work with some tinkering but
+  is not officially supported.
```
```suggestion
- A copy of Zork Grand Inquisitor on GOG. The Steam version works with some tinkering but is not officially supported.
```
Nitpicky again, just for briefness

nicholassaylor commented on worlds/zork_grand_inquisitor/docs/setup_en.md:
```diff
@@ -0,0 +1,42 @@
+- Windows OS (Hard required. Client is using memory reading / writing through Win32 API)
+- A copy of Zork Grand Inquisitor. Only the GOG version is supported. The Steam version can work with some tinkering but
+  is not officially supported.
+- ScummVM 2.7.1 64-bit (Important: Will not with any other version. [Direct Download](https://downloads.scummvm.org/frs/scummvm/2.7.1/scummvm-2.7.1-win32-x86_64.zip))
```
```suggestion
- ScummVM 2.7.1 64-bit (Important: Will not work with any other version. [Direct Download](https://downloads.scummvm.org/frs/scummvm/2.7.1/scummvm-2.7.1-win32-x86_64.zip))
```

nicholassaylor reviewed (commented):
Only reviewed docs, but looked very good there. Have a few small comments and nitpicks but is honestly fine as is.

nbrochu commented on worlds/zork_grand_inquisitor/docs/en_Zork Grand Inquisitor.md:
```diff
@@ -0,0 +1,97 @@
+## What is considered a location check in Zork Grand Inquisitor?
+
+- Solving puzzles
+- Accessing certain areas for the first time
+- Triggering certain interactions, even if they aren't puzzles per se
```
I have a full guide to all location checks linked in the same document, this is just to give a broad overview

alwaysintreble commented on worlds/zork_grand_inquisitor/game_controller.py:
```diff
@@ -0,0 +1,1447 @@
+                for key in data.statemap_keys:
+                    self._write_game_flags_value_for(key, 2)
+            else:
+                if hotspot_item == ZorkGrandInquisitorItems.HOTSPOT_666_MAILBOX:
```
kind of the same comment here. i think a lot of this can be reduced down and made more readable with dicts or maybe a dataclass.

nbrochu replied:
I disagree here. The huge if trees are absolutely ugly, no question there, but they are certainly readable (from my perspective as the person that will have to maintain this). Readability isn't just how concise the code it, it's also how quickly you can grasp what the code is meant to do. Would it be possible to abstract it to a single data structure, I would do it (like I did for the grant conditions, thanks to your suggestion), but simplifying only the parts that can be simplified implies that the logic will be split between 2 places and that actively harms readability

alwaysintreble replied:
i mean i did and it's definitely doable, but i do agree that it's likely not worth the effort and even then it wouldn't necessarily guarantee more readability due to all the specific conditions

alwaysintreble commented on worlds/zork_grand_inquisitor/game_controller.py:
```diff
@@ -0,0 +1,1447 @@
+                        self._write_game_flags_value_for(2586, 0)
+                elif hotspot_item == ZorkGrandInquisitorItems.HOTSPOT_DENTED_LOCKER:
+                    if self.game_state_manager.game_location == "th3j":
+                        five_is_open: bool = self._read_game_state_value_for(11847) == 1
```
can this be anything besides 0 or 1? i could be wrong but i'd think `bool(whatever)` would be faster. if the actual value doesn't matter, python ints are truthy so 0 will eval to false and anything else to true.

nbrochu replied:
It can. And I would rather not touch this because it doubles has self-documentation for the precise values I expect to see in RAM

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/cd6f1af6-a57b-4a1e-a1a9-2bdf9ce01364). 

I'm not a fan of simplifying without a very good reason to do so. Micro-gains in performance are not worth the obfuscation.

Don't forget that this is technically the mod side of my implementation, not the world side. This is never going to affect world loading, generation etc. Performance is not an issue at all when playing the game.

alwaysintreble commented on worlds/zork_grand_inquisitor/world.py:
```diff
@@ -0,0 +1,210 @@
+    item_name_to_item: Dict[str, ZorkGrandInquisitorItems] = item_names_to_item()
+
+    def create_regions(self) -> None:
+        deathsanity: bool = self.options.deathsanity.value == 1
```
as a rule of thumb i really don't like comparing options to arbitrary integers. some alternatives include:
```python
deathsanity: bool = bool(self.options.deathsanity)
```
```python
deathsanity: bool = self.options.deathsanity == DeathSanity.option_true
```
```python
deathsanity: bool = self.options.deathsanity == True
```
```python
deathsanity: bool = self.options.deathsanity == "true"
```

nbrochu replied:
I also didn't like that bit but wasn't sure what was possible with the new options API. I'll just use True / False then

alwaysintreble commented on worlds/zork_grand_inquisitor/world.py:
```diff
@@ -0,0 +1,210 @@
+    def create_regions(self) -> None:
+        deathsanity: bool = self.options.deathsanity.value == 1
+
+        region_mapping: Dict[ZorkGrandInquisitorRegions, Region] = dict()
+
+        region_enum_item: ZorkGrandInquisitorRegions
+        for region_enum_item in region_data.keys():
+            region_mapping[region_enum_item] = Region(region_enum_item.value, self.player, self.multiworld)
```
```suggestion
        region_enum_item: ZorkGrandInquisitorRegions
        region_mapping = {region_enum_item: Region(region_enum_item.value, self.player, self.multiworld 
                                      for region_enum_item in region_data}
```

nbrochu replied:
Not a fan of comprehensions, except when they fit on a single line. You were talking about readability in other comments, yet long comprehensions always have worse readability than loops and the whole "it's faster than a loop" has been debunked too.

I will still oblige because this is the world side of my implementation.

alwaysintreble commented on worlds/zork_grand_inquisitor/world.py:
```diff
@@ -0,0 +1,210 @@
+            region_mapping[region_enum_item] = Region(region_enum_item.value, self.player, self.multiworld)
+
+        region_locations_mapping: Dict[ZorkGrandInquisitorRegions, Set[ZorkGrandInquisitorLocations]]
+        region_locations_mapping = locations_by_region(include_deathsanity=deathsanity)
```
it's a little odd to me that you're using an unnecessary kwarg here when you don't do it elsewhere.

nbrochu replied:
Other functions that have an arg are generally explicit with their name as to what that arg is:
* `items_with_tag(tag: ZorkGrandInquisitorTags)`
* `locations_with_tag(tag: ZorkGrandInquisitorTags)`
* `location_access_rule_for(location: ZorkGrandInquisitorLocations, player: int)`

This is not the case here, hence the kwarg. Anyway, total nitpick.

alwaysintreble commented on worlds/zork_grand_inquisitor/world.py:
```diff
@@ -0,0 +1,210 @@
+
+            # Connections
+            region_exit: ZorkGrandInquisitorRegions
+            for region_exit in region_data[region_enum_item].exits or tuple():
```
what's the point of the tuple here?

nbrochu replied:
`exits` can be None. I prefer to have None in my data vs an empty tuple when there are no exits (good old null not being the same as 0)

alwaysintreble commented on worlds/zork_grand_inquisitor/world.py:
```diff
@@ -0,0 +1,210 @@
+            self.multiworld.regions.append(region)
+
+    def create_items(self) -> None:
+        quick_port_foozle: bool = self.options.quick_port_foozle.value == 1
```
same for these options

alwaysintreble commented on worlds/zork_grand_inquisitor/world.py:
```diff
@@ -0,0 +1,210 @@
+
+        total_locations: int = len(self.multiworld.get_unfilled_locations(self.player))
+
+        for _ in range(total_locations - len(item_pool)):
+            item_pool.append(self.create_item(self.get_filler_item_name()))
```
```suggestion
        item_pool += [self.create_filler() for _ in range(total_locations - len(item_pool)]
```

nbrochu replied:
Nice, didn't know there already was a helper for this. And this is a good candidate for a list comprehension too

alwaysintreble commented on worlds/zork_grand_inquisitor/world.py:
```diff
@@ -0,0 +1,210 @@
+
+        item: ZorkGrandInquisitorItems
+        data: ZorkGrandInquisitorItemData
+        for item, data in item_data.items():
```
this could all be changed into a single list comprehension too

nbrochu replied:
Not while maintaining the same level of readability. It's a fairly large block.
![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/476649a0-591c-4233-9d1a-8a7fd40d6730)

alwaysintreble commented on worlds/zork_grand_inquisitor/world.py:
```diff
@@ -0,0 +1,210 @@
+
+        region_enum_item: ZorkGrandInquisitorRegions
+        region: Region
+        for region_enum_item, region in region_mapping.items():
```
this isn't bad but if you're ever looking for a place for speedup this is a pretty good place. the way this is being done currently is relatively slow. it's probably worth doing some of the data parsing in the location class and seeing if you can format it in such a way as to call `add_locations` with this, or alternatively pre-parse the data into a comprehension that can be more easily parsed by the AP objects.

alwaysintreble commented on worlds/zork_grand_inquisitor/data_funcs.py:
```diff
@@ -0,0 +1,256 @@
+
+
+def item_names_to_item() -> Dict[str, ZorkGrandInquisitorItems]:
+    return {item.value: item for item, _ in item_data.items()}
```
```suggestion
    return {item.value: item for item in item_data}
```

nbrochu replied:
Yes, good catch

alwaysintreble commented on worlds/zork_grand_inquisitor/data_funcs.py:
```diff
@@ -0,0 +1,256 @@
+    for item, data in item_data.items():
+        if data.tags is not None:
+            for tag in data.tags:
+                groups[tag.value].add(item.value)
```
this allows you to remove the tag loop above
```suggestion
                groups.setdefault(tag.value, set()).add(item.value)
```

alwaysintreble commented on worlds/zork_grand_inquisitor/data_funcs.py:
```diff
@@ -0,0 +1,256 @@
+        ):
+            continue
+
+        mapping[data.region].add(location)
```
```suggestion
        mapping.setdefault(data.region, set()).add(location)
```

nbrochu replied:
This is not equivalent as the other place where you suggested to use `setdefault`. Menu region doesn't have locations, so it breaks. Keeping this the way it is.

alwaysintreble commented on worlds/zork_grand_inquisitor/data_funcs.py:
```diff
@@ -0,0 +1,256 @@
+        if data.tags is not None and tag in data.tags:
+            locations.add(location)
+
+    return locations
```
```suggestion
return {location for location, data in location_data.items() if data.tags is not None and tag in data.tags}
```

nbrochu replied:
Previously stated my opinion on long comprehensions

alwaysintreble commented on worlds/zork_grand_inquisitor/docs/en_Zork Grand Inquisitor.md:
```diff
@@ -0,0 +1,102 @@
+# Zork Grand Inquisitor
+
+## Where is the settings page?
```
missed this when renaming the other instances to options

alwaysintreble commented on worlds/zork_grand_inquisitor/test/test_access.py:
```diff
@@ -0,0 +1,2927 @@
+        "start_with_hotspot_items": "false",
+    }
+
+    def test_access_crossroads_to_dm_lair_sword(self) -> None:
```
tests overall are fine. worth considering combining a lot of these methods at some point because each method constructs an entirely new multiworld and calls all the steps up to pre_fill.

nbrochu replied:
Noted. But my world's entire test suite runs in ~5 seconds. If that's acceptable, I would rather not butcher them recombine them in a more optimized way that won't be as explicit to save a second.

alwaysintreble reviewed (commented):
mostly just styling and optimization comments. docs seem fine

nbrochu commented on worlds/zork_grand_inquisitor/game_controller.py:
```diff
@@ -0,0 +1,1447 @@
+            if ZorkGrandInquisitorTags.DEATHSANITY in data.tags and self.option_deathsanity is False:
+                continue
+
+            if missable_location == ZorkGrandInquisitorLocations.BOING_BOING_BOING:
```
I managed to abstract this specific one into a separate data structure and simplified `_check_for_missable_locations_to_grant` as a result.

For the other ones, as ugly as the huge if trees are, the logic of each condition doesn't have much in common with other conditions, so I don't see what I could abstract out without severely affecting readability and I'm not willing to do that. If you take the time to look at each condition, you'll see what I mean.

nbrochu commented:
@alwaysintreble Thanks for taking the time to look at this. I addressed some of the comments. Disagree with others. I think it's fair that if I am to maintain this, I can determine what has superior readability in my book, especially on the mod side (this implementation contains both the world and client-side mod).

For the speed aspect in `create_regions`, as of now, generation with the most expensive options clocks in at ~50ms (no playthrough). I don't think that's horribly slow for 114 progression items. If that's too high in the opinion of the core team, I benchmarked a few things and can probably shave off ~20ms if I pre-compute the rules and skip the evaluation on the fly from the data. It won't be pretty at all though, so I'd prefer to confirm the generation times are not in the acceptable range before committing to something like that

nbrochu commented:
I explored pre-assembling all the rule strings for locations and entrances in a python file so I don't have to assemble them on the spot during generation and saw literally no performance gain. So that code is fast enough.

Honestly, I think my 50ms with the most expensive options is more than good enough when generating with a template Stardew takes ~4s and can be taken to ~9s by using more expensive options. Shaving off single milliseconds is not worth the extra complexity I would need to add to the code base IMO

alwaysintreble reviewed (approved):
the little bits i had actual concerns about have all been addressed. fwiw i do agree that micro optimizations aren't worth reducing readability and at the end of the day it's your world so write it how you prefer as long as it's consistent. most of the original comments i made were either stray thoughts or prompts to attempt to promote thinking about approaching some of the stuff in a different manner and still then definitely not any kind of requirement though i suppose i could've worded it a bit better in that respect. pretty much all of the suggestions were intended as that and not requirements

alwaysintreble commented:
as far as rule optimizations i think the only approach that would ever really be worth it would be to have it such that each rule layout per option combination gets cached such that each world instance sharing those option combinations can share them. this is of course only relevant for larger multiworlds, and even then only really worth it if your rule evals are sufficiently slow like stardew's. you aren't doing anything egregious like checking a bunch of static information or the entire multiworld in the access rules so it's already fine

nbrochu commented:
> though i suppose i could've worded it a bit better in that respect. pretty much all of the suggestions were intended as that and not requirements

For the record, I never saw them as requirements, I just like to fully engage with suggestions when someone actually takes some of their time to share thoughts, even if I don't end up agreeing with everything. You did make me look at some stuff from a different perspective than the one I've had for months.

nbrochu commented:
> as far as rule optimizations i think the only approach that would ever really be worth it would be to have it such that each rule layout per option combination gets cached such that each world instance sharing those option combinations can share them. this is of course only relevant for larger multiworlds, and even then only really worth it if your rule evals are sufficiently slow like stardew's. you aren't doing anything egregious like checking a bunch of static information or the entire multiworld in the access rules so it's already fine

You are missing "only relevant when it's a game popular enough that it has a chance of showing up multiple times in a single multiworld" 😄

But jokes aside, I'm not sure where I'd start. Does it need to be a custom implementation? Is there a standardized way to do this? Any existing worlds I can look at to figure out a pattern?

alwaysintreble commented:
> But jokes aside, I'm not sure where I'd start. Does it need to be a custom implementation? Is there a standardized way to do this? Any existing worlds I can look at to figure out a pattern?

it'd definitely have to be custom and afaik nothing does it atm so no examples either. i considered doing it with my rewrite, but i found it a lot easier with my game to just write all the rules statically, since i wasn't able to come up with a clean solution for the difficult part of matching the player to the methods. i think it's worth thinking about and eventually coming up with some way to do it that could potentially be standardized, but i also think it might be more worth putting a lot of that effort into rewriting the rules system so that we can have proper helpers, evaluation of the rules themselves, etc., but that's also pretty far above my skill/knowledge level 😓

Silvris reviewed (approved):
Looks good, ran unittests + all games multiworld + 50 Zork gen.

50 Zork gen lasted 19.90 without playthrough (but sharing resources with Stardew tests).

EDIT: running the tests took long enough Pory caught some of these.

nbrochu commented on worlds/zork_grand_inquisitor/world.py:
```diff
@@ -0,0 +1,208 @@
+from typing import Any, Dict, List, Set, Tuple
```
I know all other worlds do everything in `__init__.py` but to me the class definitions don't belong there. `__init__.py` is responsible for the initialization of the package, not holding any of its content. There's nothing functionally wrong with how it's done in AP, it obviously works, but to me it's messy, and that should become evident the day there is interest in lazy loading worlds. Right now, there is little difference indeed, as you noted.

I had actually discussed this in the core dev channel way back, mostly with alwaysintreble, and he agreed (for what it's worth, I guess).

alwaysintreble replied:
i think i had suggested doing so 😅 i definitely like it a lot more, though i will say with the slightly abandoned lazy loading branch i have this wouldn't make any difference.

nbrochu replied:
I still maintain that lazy loading in the future or not, __init__.py should be reserved for package initialization stuff and not package content.

nbrochu commented on worlds/zork_grand_inquisitor/game_controller.py:
```diff
@@ -0,0 +1,1364 @@
+
+class GameController:
+    def __init__(self, logger=None) -> None:
+        self.logger: Optional[logging.Logger] = logger
+
+        self.game_state_manager: GameStateManager = GameStateManager()
+
+        self.received_items: Set[ZorkGrandInquisitorItems] = set()
+        self.completed_locations: Set[ZorkGrandInquisitorLocations] = set()
+
+        self.completed_locations_queue: collections.deque = collections.deque()
+        self.received_items_queue: collections.deque = collections.deque()
+
+        self.all_hotspot_items: Set[ZorkGrandInquisitorItems] = (
+            items_with_tag(ZorkGrandInquisitorTags.HOTSPOT)
+            | items_with_tag(ZorkGrandInquisitorTags.SUBWAY_DESTINATION)
+            | items_with_tag(ZorkGrandInquisitorTags.TOTEMIZER_DESTINATION)
+        )
+
+        self.game_id_to_items: Dict[int, ZorkGrandInquisitorItems] = game_id_to_items()
+
+        self.possible_inventory_items: Set[ZorkGrandInquisitorItems] = (
+            items_with_tag(ZorkGrandInquisitorTags.INVENTORY_ITEM)
+            | items_with_tag(ZorkGrandInquisitorTags.SPELL)
+            | items_with_tag(ZorkGrandInquisitorTags.TOTEM)
+        )
+
+        self.available_inventory_slots: Set[int] = set()
+
+        self.goal_completed: bool = False
+
+        self.option_goal: Optional[ZorkGrandInquisitorGoals] = None
+        self.option_deathsanity: Optional[bool] = None
+        self.option_grant_missable_location_checks: Optional[bool] = None
```
I saw that discussion. This I can certainly fix.

nbrochu commented on worlds/zork_grand_inquisitor/data_funcs.py:
```diff
@@ -0,0 +1,247 @@
+        if isinstance(requirement, tuple):
+            lambda_string += "("
+
+            ii: int
```
They are the indices coming from using `enumerate`. It's extremely common in Python to do it like this. You can find multiple examples of core doing it exactly like this.

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/78494b69-f2f5-441f-b481-a13a9280503d)

The `ii` is less common but it's how I've personally always handled 2D lists. Others prefer `i` `j` `k`. I prefer `ii` because it implies the level of nesting you are at and I can always tell it's meant to be an index at a glance

nbrochu commented on worlds/zork_grand_inquisitor/game_controller.py:
```diff
@@ -0,0 +1,1364 @@
+
+                        left_inflated: bool = (raft_in_left and raft_inflated) or (captain_in_left and captain_inflated)
+
+                        right_inflated: bool = (raft_in_right and raft_inflated) or (
+                                captain_in_right and captain_inflated
+                        )
```
My entire code base is automatically formatted with [Black](https://github.com/psf/black). It's an opiniated PEP8 formatter maintained by the Python Software Foundation themselves. I don't always like the way it formats certain things, but it's as consistent as could be.

nbrochu commented on worlds/zork_grand_inquisitor/game_controller.py:
```diff
@@ -0,0 +1,1364 @@
+
+                self._check_for_completed_locations()
+
+                if self.option_grant_missable_location_checks is True:
```
This one is more a force of habit. There is a difference in intent in checking for truthy / falsy values (especially since in Python, every type can be tested for truthiness) and specifically looking for True / False. Omitting `is True` is not just formatting shorthand.

I can change it, as it seems to bother Silvris too.

nbrochu commented on worlds/zork_grand_inquisitor/enums.py:
```diff
@@ -0,0 +1,350 @@
+import enum
```
This is mostly a stylistic choice. I prefer to always know explicitly from which other module a callable or foreign classes are from in my code. It can lead to things that seem sillier like this, but I'm super consistent about it. I'm well aware it tends to annoy people 😄

nbrochu commented on worlds/zork_grand_inquisitor/options.py:
```diff
@@ -0,0 +1,61 @@
+from Options import Choice, DefaultOnToggle, PerGameCommonOptions, Toggle
+
+
+class Goal(Choice):
+    """
+    Determines the victory condition
+
+    Three Artifacts: Retrieve the three artifacts of magic and place them in the walking castle
+    """
+    display_name: str = "Goal"
+
+    default: int = 0
+    option_three_artifacts: int = 0
```
I have different goals planned for the future, so my code is already prepared for this eventuality by always treating the goal as a variable, even if there is only a single one for now. 

I agree that the user-facing consequence of doing this will seem odd. Let me think about it

nbrochu replied:
Thought about it. I hope we are ready to accept it as a quirk (I'm fairly certain I've seen other worlds have single choice options like that to be expansion friendly). I considered removing the option and passing a hardcoded value to `GameController` instead until there is at least a second goal, but unfortunately, it's not the only usage. The PopTracker pack also makes use of it, and honestly, I would rather not revisit that if possible. There's already more than 30 hours of (unenjoyable) work that went in that pack.

I could also add a note in the option docstring, but that may be awkward too. I don't get the feeling many users will care, and I'm not a fan of over-explaining / justifying things when it's not strictly necessary (i.e. here's why this option only has 1 value for now)

nbrochu commented on worlds/zork_grand_inquisitor/data_funcs.py:
```diff
@@ -0,0 +1,247 @@
+    data: ZorkGrandInquisitorLocationData = location_data[location]
+
+    if data.requirements is None:
+        return "lambda state: True"
+
+    lambda_string: str = "lambda state: "
+
+    i: int
+    requirement: Union[
+        Tuple[
+            Union[
+                ZorkGrandInquisitorEvents,
+                ZorkGrandInquisitorItems,
+            ],
+            ...,
+        ],
+        ZorkGrandInquisitorEvents,
+        ZorkGrandInquisitorItems
+    ]
+
+    for i, requirement in enumerate(data.requirements):
+        if isinstance(requirement, tuple):
+            lambda_string += "("
+
+            ii: int
+            sub_requirement: Union[ZorkGrandInquisitorEvents, ZorkGrandInquisitorItems]
+            for ii, sub_requirement in enumerate(requirement):
+                lambda_string += f"state.has(\"{sub_requirement.value}\", {player})"
+
+                if ii < len(requirement) - 1:
+                    lambda_string += " or "
+
+            lambda_string += ")"
+        else:
+            lambda_string += f"state.has(\"{requirement.value}\", {player})"
+
+        if i < len(data.requirements) - 1:
+            lambda_string += " and "
+
+    return lambda_string
```
Context here is that I went with a bottom-up approach from pure data. My data is rich but pure in the sense that it holds nothing functional in it; only literals and enum items.

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/7ca8be5f-07d0-43d3-963d-fca1f18f257f)

The majority of the functionality of my world (both AP-side and client-side) leverages this data. It's way more useful and versatile than typical tables worlds use. For rules, I know that hand-typed lambdas are more common (and more performant than evaluating at runtime, not going to try and pretend otherwise), but I find it extremely error-prone and messy. Here, I can unit test 2 functions and have peace of mind that everything will work as intended from a simple data structure.

nbrochu replied:
That being said, it's not that I specifically WANT to do lambda strings with evaluation. I just haven't found a better way that works with the moderate level of rule complexity I have.

If someone can show me a better way to do it from the data I have, I am 100% in.

Berserker66 replied:
have you perused https://docs.python.org/3/library/ast.html yet?

nbrochu commented:
@PoryGone @Silvris

Thank you for taking some of your time to review. You can see what I've addressed in this [commit](https://github.com/ArchipelagoMW/Archipelago/pull/2539/commits/d65810018abfbea29fa6288c7eed2c36b346e269). I've also replied to every comment that didn't necessarily yield a code change. Didn't resolve those comments yet for visibility.

nbrochu commented on worlds/zork_grand_inquisitor/world.py:
```diff
@@ -0,0 +1,208 @@
+            self.multiworld.early_items[self.player][ZorkGrandInquisitorItems.ROPE.value] = 1
+            self.multiworld.early_items[self.player][ZorkGrandInquisitorItems.LANTERN.value] = 1
+
+            if start_with_hotspot_items is False:
```
I gave my reasoning [here](https://github.com/ArchipelagoMW/Archipelago/pull/2539#discussion_r1519692903) in case you are curious. That being said, the distinction is not really useful when you aren't dealing with mixed types, so I changed all usage of `is True|False`

Berserker66 commented on worlds/zork_grand_inquisitor/client.py:
```diff
@@ -0,0 +1,193 @@
+        if cmd == "Connected":
+            self.game = self.slot_info[self.slot].game
+
+        Utils.async_start(process_package(self, cmd, _args))
```
it's odd, but certainly not  an error, to have an if  cmd == connected, then  start an async task always for every packet, but that handler only reacts to cmd == Connected too.

Berserker66 replied:
since the function is doing nothing that can be awaited, I'd just do it right here.
